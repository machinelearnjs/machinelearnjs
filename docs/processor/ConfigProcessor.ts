import * as fs from 'fs-extra';
import { isEmpty } from 'lodash';
import * as path from 'path';
import { BaseProcesser } from './BaseProcesser';

export class ConfigProcessor extends BaseProcesser {
  private srcConfigPath = path.join(__dirname, '../config.ts');
  private destConfigPath = path.join(__dirname, '../md_out/.vuepress/config.ts');
  private srcPublicPath = path.join(__dirname, '../public');
  private destPublicPath = path.join(__dirname, '../md_out/.vuepress/public');
  private srcOverrideStylePath = path.join(__dirname, '../palette.scss');
  private destOverrideStylePath = path.join(__dirname, '../md_out/.vuepress/styles/palette.scss');

  /**
   * Runs the processor
   * @param {any} apiChildren
   */
  public run({ apiChildren }): void {
    if (isEmpty(apiChildren)) {
      throw Error('Cannot execute the processor because apiChildren is empty');
    }
    this.createDir();
    // 1. config
    fs.createReadStream(this.srcConfigPath).pipe(fs.createWriteStream(this.destConfigPath));

    // 2. public
    fs.copySync(this.srcPublicPath, this.destPublicPath);

    // 3. Style
    fs.mkdirSync(path.dirname(this.destOverrideStylePath), { recursive: true });
    fs.createReadStream(this.srcOverrideStylePath).pipe(fs.createWriteStream(this.destOverrideStylePath));
  }

  /**
   * Create dir if not exist, it can be abstracted to a Utils file
   */
  private createDir(): void {
    // Create public dir if it doesn't exist
    if (!fs.existsSync(this.destPublicPath)) {
      fs.mkdirSync(this.destPublicPath);
    }
  }
}
