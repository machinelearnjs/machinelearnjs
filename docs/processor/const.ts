// File's Class|Method kind
export const kindStringClass = 'Class';
export const kindStringConst = 'Constructor';
export const kindStringFunction = 'Function';
export const kindStringMethod = 'Method';
export const kindStringProperty = 'Property';
export const kindStringEnum = 'Enumeration';

export const kindNumberVariable = 32;
export const kindNumberFunction = 64;
export const kindNumberClass = 128;
export const kindNumberConstructor = 512;
export const kindNumberProperty = 1024;
export const kindNumberMethod = 2048;

// Parameter type
export const paramTypeReflection = 'reflection';
export const paramTypeIntrinsic = 'intrinsic';
export const paramTypeArray = 'array';
export const paramTypeReference = 'reference';
export const paramTypeUnion = 'union';

// Return type
export const returnTypeIntrinsic = 'intrinsic';
export const returnTypeArray = 'array';
export const returnTypeReference = 'reference';
export const returnTypeReflection = 'reflection';

// Return name
export const returnNamePromise = 'Promise';

// Tags
export const tagTypeExample = 'example';
export const tagTypeIgnore = 'ignore';

// Reference Kinds
export const refNumberTypeAlias = 2097152;
export const refNumberInterface = 256;
export const refKindInterface = 'Interface';
export const refKindTypeAlias = 'Type alias';
export const refTypeArgTypeUnion = 'union';
export const refTypeTypeParameter = 'typeParameter';
export const refTypeArgTypeIntrinsic = 'intrinsic';

// Matrix Types
export const type1DMatrix = 'Type1DMatrix';
export const type2DMatrix = 'Type2DMatrix';
export const type3DMatrix = 'Type3DMatrix';
export const type4DMatrix = 'Type4DMatrix';
