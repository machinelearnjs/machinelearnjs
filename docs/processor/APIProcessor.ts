import * as fs from 'fs';
import * as _ from 'lodash';
import * as path from 'path';
import { BaseProcesser } from './BaseProcesser';
import * as consts from './const';
const docsJson = JSON.parse(fs.readFileSync(path.join(__dirname, '../docs.json'), 'utf8'));

interface SidebarDefinition {
  children: [string, string];
  collapsable: boolean;
  title: string;
}

/**
 * Processor used to process API docs located under lib/src/
 */
export class APIProcessor extends BaseProcesser {
  public apiChildren = [];
  private vuepressExtraConfigPath = path.join(__dirname, '../md_out/.vuepress/apiExtra.json');
  private themePath = path.join(__dirname, '../themes/markdown');
  private entityPageFile = 'entity_page.hbs';
  private homePageFile = 'api_readme.hbs';
  private apiOutputPath = path.join(__dirname, '../md_out/api');
  private srcApiHomeTheme = path.join(this.themePath, this.homePageFile);
  private destApiHomePage = path.join(__dirname, '../md_out/api/README.md');
  private pathDelimeter = '.';
  // Class and Function
  private entityKindWhitelist = [consts.kindNumberClass, consts.kindNumberFunction]; // Whitelisting kinds when grabbing class or method
  private moduleNameBlackList = ['"'];

  /**
   * Run the processor
   * @param hbs
   */
  public run(hbs): void {
    // Creating required dir
    this.createDir();

    // Order API children
    this.apiChildren = this.retrieveOrderedAPIs(docsJson);
    // Construct the sidebar configs
    this.buildSidebar(this.apiChildren);
    // Create the API homepage
    this.processHomePage(hbs, this.apiChildren);

    // Process API pages
    this.processAPIEntityPage(hbs, this.apiChildren);
  }

  /**
   * Build the sidebar json for the API pages
   * @param apiChildren
   */
  private buildSidebar(apiChildren): void {
    const extraConfig = {
      apiSidebar: this.buildSidebarJSON(apiChildren),
    };
    // Writing extraConfig object as .vuepress/apiExtra.json
    fs.writeFileSync(this.vuepressExtraConfigPath, JSON.stringify(extraConfig), 'utf-8');
  }

  /**
   * Build a sidebar JSON for nested navigations
   * e.g. [{"title":"cluster","collapsable":false,"children":[["./cluster.KMeans","KMeans"]]}
   * @param apiChildren
   * @returns {Array}
   */
  private buildSidebarJSON(apiChildren): SidebarDefinition[] {
    return _.reduce(
      apiChildren,
      (sum, child) => {
        const [module, name] = child.name.split('.');
        const existingGroupIndex = _.findIndex(sum, (o) => o.text === module);
        if (existingGroupIndex === -1) {
          // If there's no existing module group according to the current child's name
          // create a new definition and append it to the sum
          const newDefinition = {
            children: [{ link: `/api/${child.name}`, text: name }],
            collapsable: true,
            text: module,
          };
          return _.concat(sum, [newDefinition]);
        } else {
          // If there's an existing module definition,
          // then append the current child's definition to the children list
          const existing = sum[existingGroupIndex];
          const newChildren = _.concat(existing.children, [{ link: `/api/${child.name}`, text: name }]);
          const updated = _.set(existing, 'children', newChildren);
          return _.set(sum, `[${existingGroupIndex}]`, updated);
        }
      },
      [],
    );
  }

  /**
   * Util funciton to clean any unwanted chars
   * @param name
   * @returns {string}
   */
  private cleanName = (name) => {
    return _.reduce(
      this.moduleNameBlackList,
      (filteredName, blackKey) => {
        return _.replace(filteredName, blackKey, '');
      },
      name,
    );
  };

  /**
   * Retrieve ordered APIs using raw docs generated from typedoc --json
   * @param docs
   * @returns {any[]}
   */
  private retrieveOrderedAPIs(docs): any {
    // Aggregate children
    const aggregatedFirstChildren = _.reduce(
      docs.children,
      (aggregation, moduleChild) => {
        // Looping the first children layer
        // Group child according the module name
        const [module, file] = moduleChild.name.split('/');
        // Clean any unwanted chars from the modulen name
        const cleanedModuleName = this.cleanName(module);
        // Grabbing each class or method of the module
        // Also it squashes the entities by moduelName.entityName e.g. preprocessing.OneHotEncoder
        const squashedEntityList = _.reduce(
          moduleChild.children,
          (entityList, entityChild) => {
            // Filter by entityKindWhitelist and skips if isIgnore comment is set
            if (this.entityKindWhitelist.indexOf(entityChild.kind) !== -1 && !this.isIgnore(entityChild)) {
              // each function or class name
              const entityName = entityChild.name;
              const fullEntityName = [cleanedModuleName, entityName].join(this.pathDelimeter);
              const newEntityChild = _.set(entityChild, 'name', fullEntityName);
              return _.concat(entityList, [newEntityChild]);
            }
            return entityList;
          },
          [],
        );
        // Filter out undefined entity appended as a result of whitelisting during the reduce
        const filteredEntityList = _.filter(squashedEntityList, (x) => !_.isUndefined(x));

        // Concat the squashedEntityList to the _.reduce aggregation
        // Also applies a shallow flatten as squashedEntityList is an array
        return _.flatten(_.concat(aggregation, filteredEntityList));
      },
      [],
    );

    // Ordering each entity by its name
    return _.orderBy(aggregatedFirstChildren, ['name']);
  }

  /**
   * Create API directory if not exist
   */
  private createDir(): void {
    // 1.2. creating the second portion: /Users/jasons/Desktop/machinelearnjs/docs/md_out/pages
    if (!fs.existsSync(this.apiOutputPath)) {
      fs.mkdirSync(this.apiOutputPath);
    }
  }

  /**
   * Process API folder's homepage aka README
   */
  private processHomePage(hbs, apiChildren): void {
    const grouped = _.groupBy(apiChildren, (o) => o.name.split(this.pathDelimeter)[0]);
    const keys = _.keys(grouped);
    const restructedChildren = _.map(keys, (key) => {
      return {
        key,
        value: _.get(grouped, key),
      };
    });

    const apiHomePageThemeContent = fs.readFileSync(this.srcApiHomeTheme, 'utf8');
    const template = hbs.compile(apiHomePageThemeContent);
    const compiledPage = template(restructedChildren);
    fs.appendFileSync(this.destApiHomePage, compiledPage, { flag: 'a' });
  }

  /**
   * Check if the passed in child has a commnet "ignore"
   * 1. If child does not have comment, search for an identical signature and pull out @ignore tag if it exists
   * @param child
   */
  private isIgnore(child): boolean {
    // Find ignores from given tags
    const findIgnores = (givenTags) =>
      _.find(givenTags, (tag) => {
        return tag.tag === consts.tagTypeIgnore;
      });

    // If child does not have comment attribute by default, then search for a signature
    if (!child.comment) {
      const { signatures } = child;
      const identSignature = _.find(signatures, (sig) => {
        return sig.name === child.name;
      });
      const foundTags = _.get(identSignature, 'comment.tags', []);
      const foundIgnores = findIgnores(foundTags);
      return !_.isEmpty(foundIgnores);
    }
    // Otherwise, evaluate using the parent comment
    const tags = _.get(child, 'comment.tags', []);
    const ignore = findIgnores(tags);
    return !_.isEmpty(ignore);
  }

  /**
   * Processes API entity pages
   * @param hbs
   * @param children
   */
  private processAPIEntityPage(hbs, children): void {
    // themes hbs files paths
    const entityPageThemePath = path.join(this.themePath, this.entityPageFile);
    const entityPageThemeContent = fs.readFileSync(entityPageThemePath, 'utf8');

    _.forEach(children, (entityChild) => {
      // 1. pages/
      // - create pages using the content
      const fullPath = path.join(this.apiOutputPath, `${entityChild.name}.md`);
      const template = hbs.compile(entityPageThemeContent);
      const compiledPage = template(entityChild);
      fs.appendFileSync(fullPath, compiledPage, { flag: 'a' });
    });
  }
}
