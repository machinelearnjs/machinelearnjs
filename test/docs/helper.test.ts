import * as fs from 'fs';
import * as path from 'path';
import {
  cleanHyperLink,
  constructMatrixType,
  constructParamTable,
  filterByKind,
  filterByTag,
  getText,
  ifEquals,
  isSignatureValid,
  renderMethodBracket,
  renderMethodReturnType,
  renderNewLine,
  renderSourceLink,
  searchInterface,
  traverseArrayDefinition,
} from '../../docs/processor';
import * as constants from '../../docs/processor/const';
import { kindNumberConstructor, kindNumberVariable } from '../../docs/processor/const';
const docsJson = JSON.parse(fs.readFileSync(path.join(__dirname, './__snapshots__/docs.snapshot.json'), 'utf8'));

/**
 * Mocking handlebar options
 */
const optionsMock = {
  fn: (x) => ({
    result: true,
    payload: x,
  }),
  inverse: (x) => ({
    result: false,
    payload: x,
  }),
};

describe('docs:helper', () => {
  describe('ifEquals', () => {
    it('should return fn when the inputs are 1 and 1', () => {
      const result = ifEquals({}, 1, 1, optionsMock).result;
      expect(result).toBe(true);
    });

    it('should return inverse when the inputs are 1 and 2', () => {
      const result = ifEquals({}, 1, 2, optionsMock).result;
      expect(result).toBe(false);
    });

    it('should return fn when the two input objects are same', () => {
      const result = ifEquals({}, { z: 1 }, { z: 1 }, optionsMock).result;
      expect(result).toBe(true);
    });

    it('should return inverse when the two input objects are not same', () => {
      const result = ifEquals({}, { z: 1 }, { z: 2 }, optionsMock).result;
      expect(result).toBe(false);
    });
  });

  describe('filterByKind', () => {
    const fakePayload = [
      // Constructor
      {
        id: 724,
        name: 'zzzz',
        kind: 512,
        kindString: 'Constructor',
        flags: {},
        comment: {
          shortText: 'model training epochs',
        },
        sources: [
          {
            fileName: 'linear_model/stochastic_gradient.ts',
            line: 126,
            character: 10,
          },
        ],
        type: {
          type: 'intrinsic',
          name: 'number',
        },
      },
      // Public var
      {
        id: 725,
        name: 'weights',
        kind: 32,
        kindString: 'Variable',
        comment: {
          shortText: 'Model training weights',
        },
        sources: [
          {
            fileName: 'linear_model/stochastic_gradient.ts',
            line: 130,
            character: 11,
          },
        ],
        type: {
          type: 'array',
          elementType: {
            type: 'intrinsic',
            name: 'number',
          },
        },
      },
    ];
    it('should filter all public Variables', () => {
      const result = filterByKind(fakePayload, optionsMock, kindNumberVariable);
      expect(result).toMatchSnapshot();
    });

    it('should filter all constructor', () => {
      const result = filterByKind(fakePayload, optionsMock, kindNumberConstructor);
      expect(result).toMatchSnapshot();
    });
  });

  describe('filterByTag', () => {
    const fakePayload = [
      {
        tag: 'example',
        text:
          "\nimport { SGDRegressor } from 'kalimdor/linear_model';\nconst reg = new SGDRegressor();\nconst X = [[0., 0.], [1., 1.]];\nconst y = [0, 1];\nreg.fit(X, y);\nreg.predict([[2., 2.]]); // result: [ 1.281828588248001 ]\n\n",
      },
      {
        tag: 'test',
        text: 'yo',
      },
    ];
    it('should find a tag example', () => {
      const result = filterByTag(fakePayload, optionsMock, 'example');
      expect(result).toMatchSnapshot();
    });

    it('should not find a tag zz', () => {
      const result = filterByTag(fakePayload, optionsMock, 'zz');
      expect(result.result).toBe(false);
    });
  });

  describe('searchInterface', () => {
    it('should find reference', () => {
      const result = searchInterface(docsJson, 1028);
      const { id, name, kind } = result;
      expect(id).toBe(1028);
      expect(name).toBe('SVMOptions');
      expect(kind).toBe(constants.refNumberInterface);
    });
    it('should invalid ID reference return null', () => {
      const result = searchInterface(docsJson, 9999999);
      expect(result).toBe(null);
    });
  });

  describe('isSignatureValid', () => {
    it('should return true for the 2nd child', () => {
      const ele = docsJson.children[0].children[0].children[0];
      const result = isSignatureValid(ele, optionsMock);
      expect(result.result).toBe(true);
      expect(result.payload).toMatchSnapshot();
    });

    it('should return false for the 1st child', () => {
      const ele = docsJson.children[0];
      const result = isSignatureValid(ele, optionsMock);
      expect(result.result).toBe(false);
    });

    it('should return false for null input', () => {
      expect(() => isSignatureValid(null, optionsMock)).toThrow(
        "Cannot read properties of null (reading 'signatures')",
      );
    });
  });

  describe('traverseArrayDefinition', () => {
    const dummy1 = {
      type: 'array',
      elementType: {
        type: 'intrinsic',
        name: 'string',
      },
    };
    const dummy2 = {
      type: 'array',
      elementType: {
        type: 'array',
        elementType: {
          type: 'intrinsic',
          name: 'number',
        },
      },
    };

    it('should dummy1 return number[]', () => {
      const result = traverseArrayDefinition(dummy1);
      expect(result).toBe('string[]');
    });

    it('should dummy2 return number[][]', () => {
      const result = traverseArrayDefinition(dummy2);
      expect(result).toBe('number[][]');
    });

    it('should throw exceptions when invalid input is given', () => {
      expect(() => traverseArrayDefinition(null)).toThrow("Cannot read properties of null (reading 'elementType')");
      expect(() => traverseArrayDefinition(123)).toThrow("Cannot read properties of undefined (reading 'name')");
    });
  });

  describe('constructMatrixType', () => {
    it('should construct Type1DMatrix into number[]', () => {
      const matrixType = constructMatrixType('Type1DMatrix', [{ name: 'number', type: 'number' }]);
      expect(matrixType).toBe('number[]');
    });
    it('should construct Type2DMatrix into number[][]', () => {
      const matrixType = constructMatrixType('Type2DMatrix', [{ name: 'number', type: 'number' }]);
      expect(matrixType).toBe('number[][]');
    });
    it('should construct Type3DMatrix into number[][][]', () => {
      const matrixType = constructMatrixType('Type3DMatrix', [{ name: 'number', type: 'number' }]);
      expect(matrixType).toBe('number[][][]');
    });
    it('should construct Type4DMatrix into number[][][]', () => {
      const matrixType = constructMatrixType('Type4DMatrix', [{ name: 'number', type: 'number' }]);
      expect(matrixType).toBe('number[][][][]');
    });
    it('should throw an error on invalid inputs', () => {
      expect(() => constructMatrixType(null, [{ name: 'number', type: 'number' }])).toThrow(
        'dim should not be null or undefined',
      );
      expect(() => constructMatrixType('Type1DMatrix', null)).toThrow('types cannot be empty!');
    });
  });

  describe('getText', () => {
    const dummy1 = {
      comment: {
        summary: [
          {
            text: 'dummy text',
            shortText: 'dummy shortText',
          },
        ],
      },
    };
    const dummy2 = {
      comment: {
        text: 'dummy legacy text',
      },
    };
    it('should dummy1 get text', () => {
      const text = getText(dummy1);
      expect(text).toBe('dummy text');
    });

    it('legacy getText', () => {
      const text = getText(dummy2);
      expect(text).toBe('dummy legacy text');
    });

    it('should throw an error for an invalid input', () => {
      expect(() => getText(null)).toThrow('Param should not be null or undefined');
      expect(() => getText(undefined)).toThrow('Param should not be null or undefined');
    });
  });

  describe('constructParamTable', () => {
    it('should build a table for param1', () => {
      const params = docsJson.children[1].children[0].children[1].signatures[0].parameters;
      const result = constructParamTable(params);
      expect(result).toMatchSnapshot();
    });

    it('should build a table for params with Type1DMatrix', () => {
      // Testing LinearRegression's fit function
      const params = docsJson.children[5].children[2].children[1].signatures[0].parameters;
      const result = constructParamTable(params);
      expect(result).toMatchSnapshot();
    });

    it('should build table for params with Type2DMatrix', () => {
      // Testing SGDClassifier's fit function
      const params = docsJson.children[5].children[6].children[5].signatures[0].parameters;
      const result = constructParamTable(params);
      expect(result).toMatchSnapshot();
    });

    it('should build table for params with Type2DMatrix of multiple types', () => {
      // Testing GaussianNB's fit function
      const params = docsJson.children[8].children[0].children[1].signatures[0].parameters;

      const result = constructParamTable(params);
      expect(result).toMatchSnapshot();
    });
  });

  describe('renderMethodReturnType', () => {
    it('should build a return type for any[]', () => {
      // Testing with imputer fit_transform
      const type = docsJson.children[10].children[1].children[2].signatures[0].type;
      const returnType = renderMethodReturnType(type);
      expect(returnType).toEqual('any[]');
    });
    it('should build a return type for void', () => {
      // Testing with KNN fit
      const type = docsJson.children[9].children[0].children[2].signatures[0].type;
      const returnType = renderMethodReturnType(type);
      expect(returnType).toEqual('void');
    });

    it('should build a toJSON return type', () => {
      // Testing with KNN toJSON
      const type = docsJson.children[9].children[0].children[4].signatures[0].type;
      const returnType = renderMethodReturnType(type);
      expect(returnType).toMatchSnapshot();
    });
  });

  describe('methodBracket', () => {
    // TODO: renderMethodBracket should decompose all the namedParameters
    it('should print a constructor parameter table', () => {
      const parameters = docsJson.children[5].children[1].children[0].signatures[0].parameters;
      const result = renderMethodBracket(parameters);
      expect(result).toEqual('(__namedParameters: *`object`*)');
    });
  });

  describe('renderSourceLink', () => {
    const source1 = [
      {
        fileName: 'ensemble/forest.ts',
        line: 80,
        character: 15,
      },
    ];
    const source2 = [
      {
        fileName: 'svm/classes.ts',
        line: 254,
        character: 34,
      },
    ];
    it('should render sources for source1', () => {
      const result = renderSourceLink(source1);
      expect(result).toEqual(
        '[ensemble/forest.ts:80](https://github.com/machinelearnjs/machinelearnjs/blob/master/src/lib/ensemble/forest.ts#L80)',
      );
    });

    it('should render sources for source2', () => {
      const result = renderSourceLink(source2);
      expect(result).toEqual(
        '[svm/classes.ts:254](https://github.com/machinelearnjs/machinelearnjs/blob/master/src/lib/svm/classes.ts#L254)',
      );
    });

    it('should not render sources for null', () => {
      expect(renderSourceLink(null)).toBe('');
      expect(renderSourceLink(123)).toBe('');
    });
  });

  describe('docs:test:renderNewLine', () => {
    it('should render a newline upon calling the method', () => {
      const result = renderNewLine();
      expect(result).toMatchSnapshot();
    });
  });

  describe('docs:test:cleanHyperLink', () => {
    it('should clean "explained_variance"', () => {
      // toJSON
      const result = cleanHyperLink('explained_variance');
      expect(result).toEqual('explained-variance');
    });

    it('should clean "toJSON"', () => {
      const result = cleanHyperLink('toJSON');
      expect(result).toEqual('tojson');
    });

    it('should not clean invalid values', () => {
      expect(() => cleanHyperLink(null)).toThrow('Should not clean values other than strings');
      expect(() => cleanHyperLink(123 as any)).toThrow('Should not clean values other than strings');
      expect(() => cleanHyperLink(undefined)).toThrow('Should not clean values other than strings');
    });
  });
});
