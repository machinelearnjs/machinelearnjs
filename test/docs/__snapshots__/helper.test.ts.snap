// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`docs:helper constructParamTable should build a table for param1 1`] = `
"| Param | Type | Default | Description |
| ------ | ------ | ------ | ------ |
| type | string |  | 
| options.delimiter | string | ',' | 
| options.lastIsTarget |  | true | 
| options.targetType | string | 'float' | 
| options.trainType | string | 'float' | 
"
`;

exports[`docs:helper constructParamTable should build a table for params with Type1DMatrix 1`] = `
"| Param | Type | Default | Description |
| ------ | ------ | ------ | ------ |
| X | number |  | training values
| y | number |  | target values
"
`;

exports[`docs:helper constructParamTable should build table for params with Type2DMatrix 1`] = `
"| Param | Type | Default | Description |
| ------ | ------ | ------ | ------ |
| X | number[][] | null | Matrix of data
| y | number[] | null | Matrix of targets
"
`;

exports[`docs:helper constructParamTable should build table for params with Type2DMatrix of multiple types 1`] = `
"| Param | Type | Default | Description |
| ------ | ------ | ------ | ------ |
| X | number[][] | null | array-like or sparse matrix of shape = [n_samples, n_features]
| y | unknown | null | array-like, shape = [n_samples] or [n_samples, n_outputs]
"
`;

exports[`docs:helper docs:test:renderNewLine should render a newline upon calling the method 1`] = `
"
"
`;

exports[`docs:helper filterByKind should filter all constructor 1`] = `
{
  "payload": [
    {
      "comment": {
        "shortText": "model training epochs",
      },
      "flags": {},
      "id": 724,
      "kind": 512,
      "kindString": "Constructor",
      "name": "zzzz",
      "sources": [
        {
          "character": 10,
          "fileName": "linear_model/stochastic_gradient.ts",
          "line": 126,
        },
      ],
      "type": {
        "name": "number",
        "type": "intrinsic",
      },
    },
  ],
  "result": true,
}
`;

exports[`docs:helper filterByKind should filter all public Variables 1`] = `
{
  "payload": [
    {
      "comment": {
        "shortText": "Model training weights",
      },
      "id": 725,
      "kind": 32,
      "kindString": "Variable",
      "name": "weights",
      "sources": [
        {
          "character": 11,
          "fileName": "linear_model/stochastic_gradient.ts",
          "line": 130,
        },
      ],
      "type": {
        "elementType": {
          "name": "number",
          "type": "intrinsic",
        },
        "type": "array",
      },
    },
  ],
  "result": true,
}
`;

exports[`docs:helper filterByTag should find a tag example 1`] = `
{
  "payload": [
    {
      "tag": "example",
      "text": "
import { SGDRegressor } from 'kalimdor/linear_model';
const reg = new SGDRegressor();
const X = [[0., 0.], [1., 1.]];
const y = [0, 1];
reg.fit(X, y);
reg.predict([[2., 2.]]); // result: [ 1.281828588248001 ]

",
    },
  ],
  "result": true,
}
`;

exports[`docs:helper isSignatureValid should return true for the 2nd child 1`] = `
{
  "flags": {
    "isExported": true,
  },
  "id": 133,
  "kind": 512,
  "kindString": "Constructor",
  "name": "constructor",
  "signatures": [
    {
      "flags": {},
      "id": 134,
      "kind": 16384,
      "kindString": "Constructor signature",
      "name": "new KMeans",
      "parameters": [
        {
          "defaultValue": " {
      distance: 'euclidean',
      k: 3,
      maxIteration: 300
    }",
          "flags": {},
          "id": 135,
          "kind": 32768,
          "kindString": "Parameter",
          "name": "options",
          "type": {
            "id": 127,
            "name": "KMeansOptions",
            "type": "reference",
          },
        },
      ],
      "type": {
        "id": 132,
        "name": "KMeans",
        "type": "reference",
      },
    },
  ],
  "sources": [
    {
      "character": 31,
      "fileName": "cluster/k_means.ts",
      "line": 45,
    },
  ],
}
`;

exports[`docs:helper renderMethodReturnType should build a toJSON return type 1`] = `
"| Param | Type | Description |
| ------ | ------ | ------ |
| classes | any[] | classes used for KNN tree
| distance | any | Choice of distance function, should choose between euclidean | manhattan
| k | number | Number of neighbors to classify
| tree | any | KNN tree
| type | string | Type of algorithm to use, choose between kdtree(default) | balltree | simple
"
`;
